File: shared/types/quiz.ts
// Shared TypeScript types for the Quiz AI System

export interface Quiz {
  id: string;
  title: string;
  description?: string;
  prompt: string;
  difficulty: 'easy' | 'medium' | 'hard';
  timeLimit: number;
  questions: Question[];
  totalQuestions: number;
  createdAt: Date;
  createdBy?: string;
  tags: string[];
}

export interface Question {
  id: string;
  text: string;
  options: {
    A: string;
    B: string;
    C: string;
    D: string;
  };
  correctAnswer: 'A' | 'B' | 'C' | 'D';
  explanation: string;
  subtopic: string;
  difficulty: 'easy' | 'medium' | 'hard';
  timeLimit?: number;
}

export interface QuizRoom {
  id: string;
  quizId: string;
  hostId: string;
  maxPlayers: number;
  status: 'waiting' | 'active' | 'finished';
  createdAt: Date;
  players: Player[];
}

export interface Player {
  id: string;
  name: string;
  sessionId: string;
  score: number;
  joinedAt: Date;
  answers: Record<string, Answer>;
  isConnected: boolean;
}

export interface Answer {
  questionId: string;
  selectedOption: 'A' | 'B' | 'C' | 'D';
  isCorrect: boolean;
  answeredAt: Date;
  timeTaken: number; // in seconds
}

export interface QuizState {
  roomId: string;
  currentQuestion: number;
  totalQuestions: number;
  timeLeft: number;
  status: 'waiting' | 'question' | 'results' | 'finished';
  questionStarted: Date;
  questionTimeLimit: number;
}

export interface WebSocketMessage {
  type: WebSocketMessageType;
  data: any;
  timestamp: Date;
  playerId?: string;
}

export type WebSocketMessageType = 
  | 'join'
  | 'leave'
  | 'player_joined'
  | 'player_left'
  | 'quiz_started'
  | 'new_question'
  | 'answer'
  | 'question_result'
  | 'leaderboard'
  | 'quiz_finished'
  | 'error'
  | 'room_deleted';

export interface Leaderboard {
  players: LeaderboardEntry[];
  roomId: string;
}

export interface LeaderboardEntry {
  playerId: string;
  name: string;
  score: number;
  correctAnswers: number;
  totalAnswers: number;
  averageTime: number;
  position: number;
}

export interface QuizGenerationRequest {
  prompt: string;
  difficulty: 'easy' | 'medium' | 'hard';
  numQuestions: number;
  timeLimit: number;
}

export interface QuizGenerationResponse {
  success: boolean;
  quiz?: Quiz;
  error?: string;
}

export interface User {
  id: string;
  email: string;
  name: string;
  createdAt: Date;
  role: 'user' | 'admin';
}

export interface AuthTokenPayload {
  userId: string;
  email: string;
  role: string;
  iat: number;
  exp: number;
}

export interface ApiResponse<T = any> {
  success: boolean;
  data?: T;
  error?: string;
  message?: string;
}

export interface PaginatedResponse<T = any> {
  data: T[];
  total: number;
  page: number;
  limit: number;
  totalPages: number;
}

// Redis Key Patterns
export const REDIS_KEYS = {
  QUIZ_ROOM: (roomId: string) => `quiz:${roomId}:room`,
  QUIZ_STATE: (roomId: string) => `quiz:${roomId}:state`,
  QUIZ_DATA: (quizId: string) => `quiz:${quizId}:data`,
  PLAYER_ANSWERS: (roomId: string, playerId: string) => `player:${roomId}:${playerId}:answers`,
  PLAYER_SCORES: (roomId: string) => `quiz:${roomId}:scores`,
  QUIZ_CHANNEL: (roomId: string) => `quiz_channel:${roomId}`,
} as const;

// Environment Variables
export interface EnvironmentConfig {
  NODE_ENV: 'development' | 'production' | 'test';
  PORT: number;
  REDIS_URL: string;
  MONGODB_URI: string;
  OPENAI_API_KEY: string;
  OPENAI_MODEL: string;
  JWT_SECRET: string;
  FRONTEND_URL: string;
  API_GATEWAY_URL: string;
  QUIZ_GENERATOR_URL: string;
  REALTIME_SERVICE_URL: string;
}

// Service URLs
export const SERVICE_URLS = {
  API_GATEWAY: process.env.API_GATEWAY_URL || 'http://localhost:3000',
  QUIZ_GENERATOR: process.env.QUIZ_GENERATOR_URL || 'http://localhost:3001',
  REALTIME_SERVICE: process.env.REALTIME_SERVICE_URL || 'http://localhost:8080',
} as const; 

File: services/socket/internal/models/quiz.go
package models

import (
	"time"
)

// LegacyQuizRoom represents a quiz room (legacy Redis-based)
type LegacyQuizRoom struct {
	ID         string    `json:"id"`
	QuizID     string    `json:"quiz_id"`
	HostID     string    `json:"host_id"`
	MaxPlayers int       `json:"max_players"`
	Status     string    `json:"status"` // waiting, active, finished
	CreatedAt  time.Time `json:"created_at"`
	Players    []Player  `json:"players"`
}

// Player represents a player in a quiz room
type Player struct {
	ID          string            `json:"id"`
	Name        string            `json:"name"`
	SessionID   string            `json:"session_id"`
	Score       int               `json:"score"`
	JoinedAt    time.Time         `json:"joined_at"`
	Answers     map[string]Answer `json:"answers"`
	IsConnected bool              `json:"is_connected"`
}

// Answer represents a player's answer to a question
type Answer struct {
	QuestionID     string    `json:"question_id"`
	SelectedOption string    `json:"selected_option"`
	IsCorrect      bool      `json:"is_correct"`
	AnsweredAt     time.Time `json:"answered_at"`
	TimeTaken      int       `json:"time_taken"` // in seconds
}

// QuizState represents the current state of a quiz
type QuizState struct {
	RoomID           string    `json:"room_id"`
	CurrentQuestion  int       `json:"current_question"`
	TotalQuestions   int       `json:"total_questions"`
	TimeLeft         int       `json:"time_left"`
	Status           string    `json:"status"` // waiting, question, results, finished
	QuestionStarted  time.Time `json:"question_started"`
	QuestionTimeLimit int      `json:"question_time_limit"`
}

// LegacyQuestion represents a quiz question (legacy Redis-based)
type LegacyQuestion struct {
	ID             string            `json:"id"`
	Text           string            `json:"text"`
	Options        map[string]string `json:"options"`
	CorrectAnswer  string            `json:"correct_answer"`
	Explanation    string            `json:"explanation"`
	TimeLimit      int               `json:"time_limit"`
	Subtopic       string            `json:"subtopic"`
	Difficulty     string            `json:"difficulty"`
}

// LegacyQuiz represents a complete quiz (legacy Redis-based)
type LegacyQuiz struct {
	ID            string           `json:"id"`
	Title         string           `json:"title"`
	Description   string           `json:"description"`
	Questions     []LegacyQuestion `json:"questions"`
	TimeLimit     int              `json:"time_limit"`
	Difficulty    string           `json:"difficulty"`
	CreatedAt     time.Time        `json:"created_at"`
	Tags          []string         `json:"tags"`
}

// WebSocketMessage represents messages sent over WebSocket
type WebSocketMessage struct {
	Type      string      `json:"type"`
	Data      interface{} `json:"data"`
	Timestamp time.Time   `json:"timestamp"`
	PlayerID  string      `json:"player_id,omitempty"`
}

// Message types
const (
	MessageTypeJoin           = "join"
	MessageTypeLeave          = "leave"
	MessageTypePlayerJoined   = "player_joined"
	MessageTypePlayerLeft     = "player_left"
	MessageTypeQuizStarted    = "quiz_started"
	MessageTypeNewQuestion    = "new_question"
	MessageTypeAnswer         = "answer"
	MessageTypeQuestionResult = "question_result"
	MessageTypeLeaderboard    = "leaderboard"
	MessageTypeQuizFinished   = "quiz_finished"
	MessageTypeError          = "error"
)

// Leaderboard represents the current standings
type Leaderboard struct {
	Players []LeaderboardEntry `json:"players"`
	RoomID  string             `json:"room_id"`
}

// LeaderboardEntry represents a player's position in the leaderboard
type LeaderboardEntry struct {
	PlayerID        string `json:"player_id"`
	Name            string `json:"name"`
	Score           int    `json:"score"`
	CorrectAnswers  int    `json:"correct_answers"`
	TotalAnswers    int    `json:"total_answers"`
	AverageTime     int    `json:"average_time"`
	Position        int    `json:"position"`
} 